FLINK-2740 Adding flink-connector-nifi module with NiFiSource and NiF…

bbende opened the pull request:
To test with a live NiFi instance, do the following:

* Download the latest 0.3.0 release: https://nifi.apache.org/download.html
* Extract it somewhere and edit conf/nifi.properties to set 

  nifi.remote.input.socket.port=9081
  nifi.remote.input.secure=false

* ./bin/nifi.sh start
* Go to http://localhost:8080/nifi/ in your browser
* Click the templates icon in the top right
* Click Browse and select the template in src/test/resources, click Import
* Drag a new template on to the canvas using the Template icon near the top left
* Press the green Start icon in the toolbar to start all processors
* Run either the NiFiSourceTopology to pull data from NiFi, or the NiFiSinkTopology to push data

zentol commented on flink-staging/flink-streaming/flink-streaming-connectors/flink-connector-nifi/src/main/java/org/apache/flink/streaming/connectors/nifi/NiFiDataPacket.java:
```diff
@@ -0,0 +1,39 @@
+
+	/**
+     * @return the contents of a NiFi FlowFile
+     */
```
you've used spaces for indentation here.

bbende replied:
good catch, checkstyles didn't find that one so I missed it, will update

zentol commented on flink-staging/flink-streaming/flink-streaming-connectors/flink-connector-nifi/src/main/java/org/apache/flink/streaming/connectors/nifi/NiFiDataPacket.java:
```diff
@@ -0,0 +1,39 @@
+	byte[] getContent();
+
+	/**
+     * @return a Map of attributes that are associated with the NiFi FlowFile
```
same as above.

zentol commented:
The NifiSource-/Topology are just examples, right? If so, could you add Example to the names? We could also unify both into a single example.

bbende commented:
I will add Example to the topology class names, do you prefer they live in src/test/resources so they don't get included in the final jar? Wasn't sure because I saw another connector had something similar in src/main/resources.

Also, I suspect the failed build is related to the template file I put in src/test/resources, I will add a license header to that file which hopefully solves that problem.

rmetzger commented:
Thank you for your contribution.

You can also check whether the build is passing by executing `mvn clean verify` locally.

Speaking about tests: is there something like a local Nifi runner which starts Nifi within the JVM?
Other systems we integrate with (HDFS, YARN, elasticsearch, kafka) have integration tests that are implemented like that. It would be great to have such a test for the Nifi connector as well.

rmetzger commented:
Which connector has the examples in `src/main/resources` ? I think its okay to put the examples into the src/main/java folder.

bbende commented:
Good to know, I'll do a full clean verify on the next update.

Regarding tests, I don't think we have plans to support running an embedded instance of NiFi for full on IT tests. We currently have a mock framework for testing processors with in NiFi, but we could potentially provide a mock implementation of the SiteToSiteClient which would help a lot for these cases, and would be useful for us in general, besides just the Flink integration.

If we were going to provide that, then we would likely need to wait until the next NiFi release (0.4.0) before finalizing this work, unless we incorporated that directly into src/test/java of the flink-connector-nifi module, and removed it later when it was generally available.

Sorry, I mistyped before and meant to say other connectors had topologies in src/main/java, twitter and rabbitmq. So it sounds like it is fine to leave the NiFi topology examples there.

rmetzger commented:
Okay, I see. I think it is fine to merge the Nifi connector without a test if there is no infrastructure for automated testing available.

rmetzger commented:
I followed the tutorial for Nifi that you posted into the pull request. Thanks a lot for the good instructions. I was able to start it and get it to run in a few minutes.

However, when starting the source topology, I got the following error:
```
10/01/2015 12:23:47	Source: Custom Source -> Map -> Sink: Unnamed(2/8) switched to RUNNING 
10/01/2015 12:23:47	Source: Custom Source -> Map -> Sink: Unnamed(4/8) switched to FAILED 
java.lang.NullPointerException
	at org.apache.flink.streaming.connectors.nifi.NiFiSource.run(NiFiSource.java:65)
	at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:57)
	at org.apache.flink.streaming.runtime.tasks.SourceStreamTask.run(SourceStreamTask.java:57)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:198)
	at org.apache.flink.runtime.taskmanager.Task.run(Task.java:579)
	at java.lang.Thread.run(Thread.java:745)
```

it seems that the `final Transaction transaction = client.createTransaction(TransferDirection.RECEIVE);` call is returning null.

This is what my NiFi screen looks like: http://i.imgur.com/FDFgrdu.png

rmetzger commented on flink-staging/flink-streaming/flink-streaming-connectors/flink-connector-nifi/src/main/java/org/apache/flink/streaming/connectors/nifi/NiFiSink.java:
```diff
@@ -0,0 +1,74 @@
+		super.close();
+		try {
+			client.close();
+		} catch (final IOException ioe) {
```
I think there is no need to catch this exception here. Flink will also log errors occurring during the close() call.

rmetzger commented on flink-staging/flink-streaming/flink-streaming-connectors/flink-connector-nifi/src/main/java/org/apache/flink/streaming/connectors/nifi/NiFiSinkTopologyExample.java:
```diff
@@ -0,0 +1,51 @@
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+package org.apache.flink.streaming.connectors.nifi;
```
can you put the two examples into a dedicated examples package?

rmetzger commented on flink-staging/flink-streaming/flink-streaming-connectors/flink-connector-nifi/src/main/java/org/apache/flink/streaming/connectors/nifi/NiFiSource.java:
```diff
@@ -0,0 +1,119 @@
+
+				// no data available. Wait a bit and try again
+				try {
+					Thread.sleep(1000L);
```
I think we should make this configurable for users with low latency requirements, ...

rmetzger commented on flink-staging/flink-streaming/flink-streaming-connectors/flink-connector-nifi/src/main/java/org/apache/flink/streaming/connectors/nifi/NiFiSource.java:
```diff
@@ -0,0 +1,119 @@
+			}
+
+			transaction.complete();
+		}
```
you need to close the `SourceContext` after the while loop

rmetzger commented on flink-staging/flink-streaming/flink-streaming-connectors/flink-connector-nifi/src/main/java/org/apache/flink/streaming/connectors/nifi/NiFiSource.java:
```diff
@@ -0,0 +1,119 @@
+	@Override
+	public void run(SourceContext<NiFiDataPacket> ctx) throws Exception {
+		while (running) {
+			final Transaction transaction = client.createTransaction(TransferDirection.RECEIVE);
```
the trxn returned here can be null.

rmetzger commented on flink-staging/flink-streaming/flink-streaming-connectors/flink-connector-nifi/src/main/java/org/apache/flink/streaming/connectors/nifi/NiFiSource.java:
```diff
@@ -0,0 +1,119 @@
+		try {
+			client.close();
+		} catch (final IOException ioe) {
+			LOG.error("Unable to close SiteToSiteClient: " + ioe.getMessage());
```
Same here. Flink can handle the Exception

rmetzger commented on flink-staging/flink-streaming/flink-streaming-connectors/flink-connector-nifi/src/main/java/org/apache/flink/streaming/connectors/nifi/NiFiSourceTopologyExample.java:
```diff
@@ -0,0 +1,55 @@
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+package org.apache.flink.streaming.connectors.nifi;
```
examples package

rmetzger commented on flink-staging/flink-streaming/flink-streaming-connectors/flink-connector-nifi/pom.xml:
```diff
@@ -0,0 +1,94 @@
+            <version>${project.version}</version>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
```
Can you clean up the dependencies section a bit to use the same indentation as the other poms?
We are using tabs instead of spaces.

bbende commented:
Working on updates to address the feedback... Regarding the NullPointerException, I'm going to protect against that in the code, but I'm still trying to think of why you ran into that. 

Did you set nifi.remote.input.socket.port in nifi.properties? 
I was able to produce that NPE when I did not set this property. 

Could you also see if any errors are showing up in  NIFI_HOME/logs/nifi-app.log at the same time?

rmetzger commented:
Thank you for looking into it. 
Your analysis was right: I've accidentally put the port into the host parameter `nifi.remote.input.socket.host=9081`. 
After I've changed it, everything was working: https://www.youtube.com/watch?v=ylnv5p5Ddk8

In my local setup, it seems that only one of my 8 parallel instances of the source are actually reading data. I suspect that Nifi is delivering data only to one reader.

How would that work in a cluster? How does Nifi know that all the Flink `NifiSource`s connecting to Nifi belong to the same system?

bbende commented:
I think the example data flow in NiFi is producing data fairly slowly, this is based on the scheduling tab of the GenerateFlowFile processor being set to 2 secs. We could set this to 0 sec to start really going by stopping the processor, right-clicking and selecting Configure. 

The SiteToSiteClientConfig has options to set batch count but the example isn't setting it so it is using some default, and the first instance of the source probably runs and pulls all the available data, and when the second instance runs nothing is left.

If you set the batch count lower in the example, like this:

    SiteToSiteClientConfig clientConfig = new SiteToSiteClient.Builder()
				.url("http://localhost:8080/nifi")
				.portName("Data for Flink")
				.requestBatchCount(5)
				.buildConfig();

and then have more than 5 flow files queued in NiFi when you start the source topology, I think the other instances will start pulling as well. This seemed to work for me using parallelism of 2, but of course let me know if you are seeing something different.

bbende commented:
I realized I didn't fully answer your question about how that works in a cluster... the SiteToSiteClient knows about all the nodes in the NiFi cluster and will pull from the output port on each node. Same concept on the sending side, it will distribute the data to different nodes in the cluster, sending more data to nodes that are considered less busy. The SiteToSiteClient is what we use internally for two NiFi instances/clusters to communicate with each other. 

Regarding how they know they belong to the same system, technically they don't, and other non-Flink clients could pull from that same port, but we have to hope that if an organization is using Flink and NiFi they would be ensuring that port was only being used by this specific Flink streaming process, and there could be multiple output ports to support multiple Flink streaming processes pulling.

There is also the option to secure a NiFi instance with SSL, and then provide SSL credentials to the SiteToSite client in order lockdown NiFi so that only authorized clients can pull/push data.

bbende commented:
Each Flink streaming job/process would pull from a different port, the port name is specified in the SiteToSiteClientConfig for the source:

![nifi-flink-multi-ports](https://cloud.githubusercontent.com/assets/605416/10226511/d819d16c-6835-11e5-8201-b378192cc8ab.png)

rmetzger commented:
Great. Thank you for your answers. I  think its fine to let users deal with cases when multiple systems want to consume data from a NiFi flow.

Once my in-code comments are addressed, the code is good to merge in my opinion.

rmetzger commented:
Ah, you've addressed my feedback. I should refresh my browser before commenting ;)

So +1, good to merge

rmetzger commented:
I'm going to merge this PR ...
